### PR TITLE
Tooling updates, exception rework.

### PR DESCRIPTION
**Pull request by sherbang**
Here's a bunch of tooling updates that encourage good coding practices.  I use these as standard.  I've broken them out into separate commits to make the results of each easier to see.

Some highlights:
- setup.py/requirements.txt replaced by Poetry which manages virtualenv for you, updating dependencies, and can publish to pypi
- pre-commit configured to do a lot of automated cleanup:
  - black autoformatting (I noticed you're already using this)
  - pyupgrade to upgrade outdated python syntax
  - ruff which is a linter based on the older pylint, with a lot of code quality checks enabled.  Some kinds of errors will be auto-fixed by ruff.

I recommend adding a plugin for ruff to your IDE.  If you use VSCode then it will be recommended to you automatically.

I also recommend github build actions to automatically run pytest and pre-commit on any PRs, and require them to pass before merging (saves a lot of time on code reviews).  I'm sure release tooling for github also exists for poetry.  I usually use Gitlab, so I'm more familiar with how to set up these things for Gitlab's CI.

Lastly this also includes some more rework of the exception handling code.  These changes ensure that important error information from the API gets passed up to the caller.

Please let me know what you think.  If you're not comfortable with any of these changes then let me know which ones and I'll resubmit.

### CODE REVIEW

**Comment by Yoshify** [comment]
I only have positive feedback about this PR! A lot of it is chores I was intending to get around to but have yet to - so I greatly appreciate this. The increased detail in exceptions is a nice touch as well.

After getting to know Poetry a bit over the last few days, I think the move to it for development is the correct decision.

Happy to merge.

Let's get in touch - how would be best to communicate with you? I've got both Slack and Discord.

File: .gitignore [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
-# Edit at https://www.toptal.com/developers/gitignore?templates=python
+## Project specific gitignore entries go up front to make them easy to find:
 
-### Python ###
+# Ignore JSON Spec files that belong to ConnectWise
+ConnectWiseManage.json
+/All.json
+/automate_json/
+
+# Are these still needed?
```
Comment (Yoshify) [inline .gitignore:8]: No longer needed - but already fixed, so I won't bother you with updating this :)

**Review by Yoshify — APPROVED** [review]
LGTM!

**Comment by sherbang** [comment]
I sent you an email from my Gozynta address. 

You can also try reaching me in discord, my username there is also sherbang.

**Pull request merged**